features/nvstore/TESTS/nvstore/functionality/main.cpp

```diff
@@ -378,6 +377,39 @@ static void nvstore_basic_functionality_test()
     delete[] nvstore_testing_buf_get;
 }
 
+// This function calculates the stack size that needs to be allocated per thread in
+// the multi-thread tests. Given minimal and maximal stack sizes, and based on the heap
+// stats (automatically enabled in CI), the function checks whether each thread has at least
+// the minimal stack size, otherwise it reduces the number of threads (something that may happen
+// on low memory boards).
+static void calc_thread_stack_size(int &num_threads, uint32_t min_size, uint32_t max_size,
+                                   uint32_t &stack_size)
+{
+    mbed_stats_heap_t heap_stats;
+    mbed_stats_heap_get(&heap_stats);
+
+    // reserved size (along with all other fields in heap stats) will be zero if
+    // app is compiled without heap stats (typically local builds)
+    if (!heap_stats.reserved_size) {
+        stack_size = max_size;
+        printf("Heap stats disabled in this build, so test may fail due to insufficient heap size\n");
+        printf("If this happens, please build the test with heap stats enabled (-DMBED_HEAP_STATS_ENABLED=1)\n");
+        return;
+    }
+
+    // Check if we can allocate enough stack size (per thread) for the current number of threads
+    while (num_threads) {
+        stack_size = (heap_stats.reserved_size - heap_stats.current_size) / num_threads;
```

Comment:
It could happen that we use whole free heap memory for threads, fallback memory should be applied
e.g. (10% of reserved_size  or 1kB)
`stack_size = (heap_stats.reserved_size - heap_stats.current_size - std::max(heap_stats.reserved_size/10, 1024);) / num_threads;`

Reply:
Understood. Are you saying this as a general note or specifically for this test? Reason I ask is that both min and max values in this test (768/2048) take these spares into account. However, I may create a more general infrastructure for this, as I get more and more low memory boards that fail on these kind of tests, so may take this into account.

Reply:
It's general note. How does it take these spares into account?
e.g. If we have 8kB free heap memory , then whole 8kB will be utilized for threads, am I right?

Reply:
Sorry, now I understand you comment. Thought the spares you mentioned were to be used by the threads themselves and not by the main thread. Yes, you have a point here.